pkg/csi/service/vanilla/controller.go

```diff
@@ -2274,6 +2267,291 @@ func (c *controller) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRe
 	return resp, err
 }
 
+func (c *controller) listSnapshotsUtilVanilla(ctx context.Context, volumeID string, snapshotID string,
+	token string, maxEntries int64) ([]*csi.Snapshot, string, error) {
+	log := logger.GetLogger(ctx)
+	var (
+		vCenterManager cnsvsphere.VirtualCenterManager
+		volManager     cnsvolume.Manager
+		vCenterHost    string
+		err            error
+	)
+	if snapshotID != "" {
+		// Retrieve specific snapshot information.
+		// The snapshotID is of format: <volume-id>+<snapshot-id>
+		volID, snapID, err := common.ParseCSISnapshotID(snapshotID)
+		if err != nil {
+			log.Errorf("Unable to determine the volume-id and snapshot-id")
+			return nil, "", err
+		}
+		// Fetch vCenterHost & volumeManager for given volume, based on VC configuration
+		vCenterManager = getVCenterManagerForVCenter(ctx, c)
+		vCenterHost, volManager, err = getVCenterAndVolumeManagerForVolumeID(ctx, c, volID, volumeInfoService)
+		if err != nil {
+			return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to get vCenter/volume manager for volume Id: %q. Error: %v", volID, err)
+		}
+		// Check for snapshot support
+		isCnsSnapshotSupported, err := vCenterManager.IsCnsSnapshotSupported(ctx, vCenterHost)
+		if err != nil {
+			return nil, "", logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to check if cns snapshot is supported on VC %s due to error: %v", vCenterHost, err)
+		}
+		if !isCnsSnapshotSupported {
+			return nil, "", logger.LogNewErrorCodef(log, codes.Unimplemented,
+				"VC %s version does not support snapshot operations", vCenterHost)
+		}
+		// Note: Re-using original Query Snapshot function since we expect only single entry for this case.
+		//		No need to handle pagination here
+		snapshots, err := common.QueryVolumeSnapshot(ctx, volManager, volID, snapID, maxEntries)
+		if err != nil {
+			// Failed to retrieve information of snapshot.
+			log.Errorf("failed to retrieve snapshot information for volume-id: %s snapshot-id: %s from vCenter %s err: %+v",
+				volID, snapID, vCenterHost, err)
+			return nil, "", err
+		}
+		return snapshots, "", nil
```

Comment:
I see that this block and the else if block is mostly same as the one in ListSnapshotUtil, can we avoid duplicate code, please maintain one set of code.

Reply:
removed duplicate code for cases where either volID or snapID is non-null